build-support/native-image/generate-native-image-for-pants-targets.bash

```diff
@@ -0,0 +1,205 @@
+#!/usr/bin/env bash
+# Script to generate a platform-specific native-image of the pantsbuild zinc wrapper which works to
+# compile macros, from a pants project.
+
+# Requires `jq` and `go`. `jq` can be downloaded from https://stedolan.github.io/jq/!
+# Works on OSX and Linux -- see the README.md in this directory for more usage info.
+
+# TODO: This will be made automatic in pants via https://github.com/pantsbuild/pants/pull/6893.
+
+set -euxo pipefail
+
+# TODO: Right now, NATIVE_IMAGE_EXTRA_ARGS='-H:IncludeResourceBundles=org.scalactic.ScalacticBundle'
+# is necessary to build any zinc native-image which compiles any code using scalatest. This
+# information should be inferred by the native-image agent, but isn't yet.
+
+# TODO: build off of a more recent graal sha (more recent ones fail to build scalac and more): see
+# https://github.com/oracle/graal/issues/1448.
+
+
+# ARGUMENTS
+
+# $@: Forwarded to a `./pants compile` invocation which runs with reflection tracing enabled.
+# $NATIVE_IMAGE_EXTRA_ARGS: Forwarded to a `native-image` invocation.
+
+
+# CONSTANTS
+
+SCRIPT_DIR="$(dirname "${BASH_SOURCE[0]}")"
+
+# shellcheck source=build-support/native-image/build-zinc-native-image.bash
+source "${SCRIPT_DIR}/build-zinc-native-image.bash"
+
+GENERATED_CONFIG_DIR="$(normalize_path "${GENERATED_CONFIG_DIR:-generated-reflect-config}")"
+GENERATED_CONFIG_JSON_FILE="${GENERATED_CONFIG_DIR}/reflect-config.json"
+GENERATED_BUILD_FILE="${GENERATED_CONFIG_DIR}/BUILD"
+
+
+# FUNCTIONS
+
+function download_buildozer {
+  >&2 go get github.com/bazelbuild/buildtools/buildozer || return "$?"
+  which buildozer
+}
+
+function build_agent_dylib {
+  pushd "$(get_substratevm_dir)" >&2
+  if is_osx; then
+    desired_agent_dylib_path="$(pwd)/libnative-image-agent.dylib"
+  else
+    desired_agent_dylib_path="$(pwd)/libnative-image-agent.so"
+  fi
+  if [[ ! -f "$desired_agent_dylib_path" ]]; then
+    mx build >&2 || return "$?"
+    mx native-image --tool:native-image-agent --verbose >&2 || return "$?"
+    if is_osx; then
+      cp -v {,lib}native-image-agent.dylib || return "$?"
+    else
+      cp -v {,lib}native-image-agent.so || return "$?"
+    fi >&2 || return "$?"
+  fi
+  echo "$desired_agent_dylib_path"
+  popd >&2
+}
+
+# run with tracing on
+
+function run_zinc_compile_with_tracing {
+  agent_lib_path="$(build_agent_dylib)"
+  if [[ ! -f "$GENERATED_CONFIG_JSON_FILE" ]]; then
+    rm -rfv "$GENERATED_CONFIG_DIR"
+    mkdir -v "$GENERATED_CONFIG_DIR"
+    # TODO: jvm options are ignored from the command line, this is a bug and should be fixed.
+    export PANTS_COMPILE_ZINC_JVM_OPTIONS="+['-agentlib:native-image-agent=config-merge-dir=${GENERATED_CONFIG_DIR}']"
+    # NB: --worker-count=1 is because the `config-merge-dir` option to the native-image agent will
+    # clobber symbols from concurrent compiles.
+    MODE=debug ./pants -ldebug \
```

Comment:
Running this in `MODE=debug` has the potential to make it take longer... would run without this.